### DISCUSSION TITLE
Guidance: Supporting both PHP 7.3 and PHP 8

### DISCUSSION THREAD

**Discussion by evansims** [lcobucci/jwt#612] [Q&A] [answered]
Good morning! 👋 I'm a developer on Auth0's PHP SDK that makes use of this library. The work being done here is tremendous and appreciated, as always!

I'm currently in the process of updating our SDK with PHP 8 support, but I'm hitting a wall as far relating to our use of this library.

The 4.X branch looks great and I'd love to adopt it, but with this branch only supporting 7.4+, it conflicts with our needs. As 7.3 is still receiving security patches for another year and a large chunk of our user base is still using it, we can't justify phasing out 7.3 support yet.

I noticed PRs fixing compatibility issues with the 3.X branch for PHP 8 have been declined, which is admittedly disappointing to see. I would contend this should be considered a bugfix and applied to the branch.

Currently, as far as I'm able to determine, this leaves me with having to fork either the 3.X branch and patch in PHP 8 support, or the 4.X branch and dropping the minimum PHP version in the composer file. Neither route is ideal.

Are there any alternatives I've overlooked here? Any suggestions on how we could handle this better? Thanks in advance

**Comment by lcobucci** [comment, marked as answer]
Hello 👋
Glad to hear your appreciation for the new API. 

So, PHP 8.0 support won't be offered on v3.x to avoid having to maintain compatibility with 3 major PHP versions and their quirks.

v3.4 provides a forward compatibility layer for v4.x. That has been done exactly to allow downstream libraries to be able to offer support for PHP 5.6+.

To achieve that you should require both versions (`^3.4 || ^4.0`) and use the components provided by v4.0 (configuration object, new token methods, and new validation API).

That approach has been used by several other libraries (like PHPUnit) and worked quite well for a few packages - such as [oauth2-server](https://github.com/thephpleague/oauth2-server/pull/1146) (maintained by @Sephster), [passport](https://github.com/laravel/passport/pull/1373) (maintained by @driesvints), and [vonage-php-sdk](https://github.com/Vonage/vonage-php-sdk-core/pull/269) (maintained by @dragonmantank)

Our [upgrading guide](https://lcobucci-jwt.readthedocs.io/en/stable/upgrading/) should provide information on how to upgrade to the new API.

Really hope this helps 🙂

**Reply by evansims** [reply]
Thanks for clarifying @lcobucci and pointing me in the right direction here. I'll investigate that route!
